### PR TITLE
`QRDecomposition` Gram-Schmidt algorithm

### PR DESCRIPTION
**Pull request by FiIipson**
Please review the code. I hope that algorithm works (it passed all the tests) however I'd be glad to resolve any issue.

### CODE REVIEW

**Comment by FiIipson** [comment]
At this moment I don't know how to deal with making the class as private as it could be and still test it.

**Review by Pankin610 — CHANGES_REQUESTED** [review]
It would be nice to have a QrDecomposition interface that would be implemented by this class (With GramScmidt), especially considering that we will make different QR implementations

File: java/src/main/java/linalg/algorithms/QRDecompositionGramSchmidt.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        norm = Math.sqrt(norm);
+
+        for (int j = 0; j < builder.VectorSize(); j++) {
+            builder.SetValue(j, builder.GetValue(j) / norm);
+        }
+
+        return builder;
+    }
+
+    double ScalarProduct(Vector a, Vector b) {
```
Comment (Pankin610) [inline java/src/main/java/linalg/algorithms/QRDecompositionGramSchmidt.java:85]: This is already implemented in VectorUtils, we shouldn't write the same thing twice

File: java/src/main/java/linalg/algorithms/QRDecompositionGramSchmidt.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            throw new IllegalArgumentException("Vectors of unequal sizes");
+        }
+
+        AtomicReference<Double> result = new AtomicReference<>((double) 0);
+        a.ForEachEntry(entry -> result.updateAndGet(v -> v + entry.Value() * b.ValueAt(entry.Index())));
+
+        return result.get();
+    }
+
+    Matrix Transpose(Matrix a) {
```
Comment (Pankin610) [inline java/src/main/java/linalg/algorithms/QRDecompositionGramSchmidt.java:96]: As previously discussed we decided to replace this with separate Matrix.Transpose method

**Review by Pankin610 — CHANGES_REQUESTED** [review]

**Comment by Pankin610** [comment]
Good job, thanks for fixing the issues

**Review by Pankin610 — APPROVED** [review]

**Pull request merged**